## Concatenar vinos blancos con tintos

In [4]:
import pandas as pd
red_wine = pd.read_csv("../../datasets/wine/winequality-red.csv", sep=";")
red_wine.head()


fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

In [9]:
red_wine.columns.values

array(['fixed acidity', 'volatile acidity', 'citric acid',
       'residual sugar', 'chlorides', 'free sulfur dioxide',
       'total sulfur dioxide', 'density', 'pH', 'sulphates', 'alcohol',
       'quality'], dtype=object)

In [12]:
white_wine = pd.read_csv("../../datasets/wine/winequality-white.csv", sep = ";")
red_wine = pd.read_csv("../../datasets/wine/winequality-red.csv", sep=";")
red_wine.head()

fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5

In [14]:
white_wine.columns.values

array(['fixed acidity', 'volatile acidity', 'citric acid',
       'residual sugar', 'chlorides', 'free sulfur dioxide',
       'total sulfur dioxide', 'density', 'pH', 'sulphates', 'alcohol',
       'quality'], dtype=object)

En python tenemos dos tipos de ejes:
- axis = 0 eje horizontal
- axis = 1 eje vertical
Vamos a concatenar estos dos dataframes uno detras de otro sin añadir nada


In [17]:
wine = pd.concat([red_wine,white_wine], axis = 0)

## Datos distribuidos
En eldirectorio distributed data tenemos +300 ficheros con cuatro columnas con la calidad del aire. Los datos ademas vienen muy distribuidos asi que la idea es concatenarlos para manejar un único dataset.

In [21]:
aire1 = pd.read_csv("../../datasets/distributed-data/001.csv")
aire1.head()

Date  sulfate  nitrate  ID
0  2003-01-01      NaN      NaN   1
1  2003-01-02      NaN      NaN   1
2  2003-01-03      NaN      NaN   1
3  2003-01-04      NaN      NaN   1
4  2003-01-05      NaN      NaN   1

In [22]:
aire1.describe()

sulfate     nitrate      ID
count  117.000000  122.000000  1461.0
mean     3.880701    0.549910     1.0
std      2.831156    0.380921     0.0
min      0.613000    0.118000     1.0
25%      2.210000    0.283500     1.0
50%      2.870000    0.453000     1.0
75%      4.730000    0.663500     1.0
max     19.100000    1.830000     1.0

Fijandonos en el count vemos que de los 1461 filas sólo117 tienen datos de sulfate y solo 112 de nitrate: es una matriz sparse con datos muy espaciados entre ellos.
La idea es usar aire1 como base y concatenar detras los ficheros uno a uno

In [31]:
filepath = "../../datasets/distributed-data/"
aire = pd.read_csv("../../datasets/distributed-data/001.csv")
final_lenght = len (aire)

for i in range (2, 333):
    s = "000{}".format(i)
    s =  filepath + s[-3:] + ".csv"
    temp = pd.read_csv(s)
    final_lenght += len(temp)
    aire = pd.concat([aire, temp], axis = 0)
print (final_lenght)
aire1.describe()


772087


sulfate        nitrate             ID
count  118783.000000  114349.000000  772087.000000
mean        3.189369       1.702932     164.461290
std         2.893127       2.525040      95.481235
min         0.000000       0.000000       1.000000
25%         1.310000       0.412000      79.000000
50%         2.370000       0.821000     168.000000
75%         4.050000       1.950000     247.000000
max        35.900000      53.900000     332.000000

## Joins de datasets
Basicamente es igual a los join de bases de datos. 

### inner join

In [35]:
filepath = "../../datasets/athletes/"
data_main = pd.read_csv(filepath+"Medals.csv", encoding = "ISO-8859-1" )
data_main.head()

Athlete   Age  Year Closing Ceremony Date  Gold Medals  \
0    Michael Phelps  23.0  2008            08/24/2008            8   
1    Michael Phelps  19.0  2004            08/29/2004            6   
2    Michael Phelps  27.0  2012            08/12/2012            4   
3  Natalie Coughlin  25.0  2008            08/24/2008            1   
4     Aleksey Nemov  24.0  2000            10/01/2000            2   

   Silver Medals  Bronze Medals  Total Medals  
0              0              0             8  
1              0              2             8  
2              2              0             6  
3              2              3             6  
4              1              3             6

In [45]:
data_main["Athlete"].unique()

array(['Michael Phelps', 'Natalie Coughlin', 'Aleksey Nemov', ...,
       'Wietse van Alten', 'Sandra Wagner-Sachse', 'Rod White'],
      dtype=object)

In [46]:
data_country = pd.read_csv(filepath+"Athelete_Country_Map.csv",encoding = "ISO-8859-1"  )
data_country.describe()

Athlete        Country
count             6968           6970
unique            6955            110
top     Slobodan Nikic  United States
freq                 2            878

In [48]:
data_sport = pd.read_csv(filepath+"Athelete_Sports_Map.csv",encoding = "ISO-8859-1"  )
data_sport.describe()

Athlete      Sport
count      6973       6975
unique     6955         49
top     Juliana  Athletics
freq          2        568

### joins con pd.merge

In [53]:
data_country = data_country.drop_duplicates(subset="Athlete")

In [54]:
data_main_country = pd.merge(left = data_main, right = data_country, left_on="Athlete", right_on="Athlete" )

data_main_country.head()

Athlete   Age  Year Closing Ceremony Date  Gold Medals  \
0    Michael Phelps  23.0  2008            08/24/2008            8   
1    Michael Phelps  19.0  2004            08/29/2004            6   
2    Michael Phelps  27.0  2012            08/12/2012            4   
3  Natalie Coughlin  25.0  2008            08/24/2008            1   
4  Natalie Coughlin  21.0  2004            08/29/2004            2   

   Silver Medals  Bronze Medals  Total Medals        Country  
0              0              0             8  United States  
1              0              2             8  United States  
2              2              0             6  United States  
3              2              3             6  United States  
4              2              1             5  United States

In [55]:
data_sport = data_sport.drop_duplicates(subset="Athlete")

In [58]:
data_all = pd.merge(left = data_main_country, right = data_sport, left_on="Athlete", right_on="Athlete" )
data_all.head(25)

Athlete   Age  Year Closing Ceremony Date  Gold Medals  \
0     Michael Phelps  23.0  2008            08/24/2008            8   
1     Michael Phelps  19.0  2004            08/29/2004            6   
2     Michael Phelps  27.0  2012            08/12/2012            4   
3   Natalie Coughlin  25.0  2008            08/24/2008            1   
4   Natalie Coughlin  21.0  2004            08/29/2004            2   
5   Natalie Coughlin  29.0  2012            08/12/2012            0   
6      Aleksey Nemov  24.0  2000            10/01/2000            2   
7      Alicia Coutts  24.0  2012            08/12/2012            1   
8     Missy Franklin  17.0  2012            08/12/2012            4   
9        Ryan Lochte  27.0  2012            08/12/2012            2   
10       Ryan Lochte  24.0  2008            08/24/2008            2   
11       Ryan Lochte  20.0  2004            08/29/2004            1   
12   Allison Schmitt  22.0  2012            08/12/2012            3   
13   Allison Schmitt  18.0  2008            08/24/2008            0   
14        Ian Thorpe  17.0  2000            10/01/2000            3   
15        Ian Thorpe  21.0  2004            08/29/2004            2   
16       Dara Torres  33.0  2000            10/01/2000            2   
17       Dara Torres  41.0  2008            08/24/2008            0   
18     Cindy Klassen  26.0  2006            02/26/2006            1   
19     Cindy Klassen  22.0  2002            02/24/2002            0   
20     Nastia Liukin  18.0  2008            08/24/2008            1   
21     Marit Bjørgen  29.0  2010            02/28/2010            3   
22     Marit Bjørgen  25.0  2006            02/26/2006            0   
23     Marit Bjørgen  21.0  2002            02/24/2002            0   
24          Sun Yang  20.0  2012            08/12/2012            2   

    Silver Medals  Bronze Medals  Total Medals        Country  \
0               0              0             8  United States   
1               0              2             8  United States   
2               2              0             6  United States   
3               2              3             6  United States   
4               2              1             5  United States   
5               0              1             1  United States   
6               1              3             6         Russia   
7               3              1             5      Australia   
8               0              1             5  United States   
9               2              1             5  United States   
10              0              2             4  United States   
11              1              0             2  United States   
12              1              1             5  United States   
13              0              1             1  United States   
14              2              0             5      Australia   
15              1              1             4      Australia   
16              0              3             5  United States   
17              3              0             3  United States   
18              2              2             5         Canada   
19              0              1             1         Canada   
20              3              1             5  United States   
21              1              1             5         Norway   
22              1              0             1         Norway   
23              1              0             1         Norway   
24              1              1             4          China   

                   Sport  
0               Swimming  
1               Swimming  
2               Swimming  
3               Swimming  
4               Swimming  
5               Swimming  
6             Gymnastics  
7               Swimming  
8               Swimming  
9               Swimming  
10              Swimming  
11              Swimming  
12              Swimming  
13              Swimming  
14              Swimming  
15              Swimming  
16              